### DISCUSSION TITLE
C#: Does swig support typemaps with more than one parameter?

### DISCUSSION THREAD

**Discussion by GA02HE** [swig/swig#2294] [Q&A] [answered]
Hello,

I would like to wrap a function for C# which takes two parameters, for example, this is the C++ function I want to wrap:

    int FindClosestPoint (const Point3D iOtherPoints[], const int iNbPts) const;

However, specifying "numinputs=2" in the typemap gives the following error:

    error : Multiple-input typemaps (numinputs > 1) not supported by this target language module.

Below is the typemap I am using:

%typemap(ctype) const Point3D iOtherPoints[] "Point3D**";
%typemap(in, numinputs=2) (const Point3D iOtherPoints[], const int iNbPts)
%{
  $1 = new Point3D[$2];
  for (int i = 0; i < $2; ++i)
  {
    $1[i] = Point3D(*$input[i]);
  }
%}
%typemap(freearg) (const Point3D iOtherPoints[], const int iNbPts)
{
  if ($1)
  {
    delete[] $1;
    $1 = nullptr;
  }
}

If I don't specify "numinputs=2", the second parameter disappears in the C++ wrapper function.

How to correctly wrap this kind of functions? Can this be done using Swig?

Thanks.

**Comment by ojwb** [comment, marked as answer]
Looking at the source code, it seems `numinputs` of more than 1 isn't supported for **any** target language.  This error is gated by:

```
Source/Modules/lang.cxx:        if (!multiinput && (GetInt(k, "value") > 1)) {
```

Looking at where `multiinput` gets set:

```
$ git grep multiinput
Source/Modules/lang.cxx:multiinput(0),
Source/Modules/lang.cxx:        if (!multiinput && (GetInt(k, "value") > 1)) {
Source/Modules/lang.cxx:  multiinput = val;
Source/Modules/swigmod.h:  int multiinput;
```

So the class constructor initialises it to zero, and there's a single place that assigns, which is in `Language::allow_multiple_input()`, which it seems is never called:

```
$ git grep allow_multiple_input
Source/Modules/lang.cxx: * Language::allow_multiple_input()
Source/Modules/lang.cxx:void Language::allow_multiple_input(int val) {
Source/Modules/swigmod.h:  void allow_multiple_input(int val = 1);
```

Looking at `git log -S` output it seems to have been this way since at least 2002.  The earlier history seems rather messed up (probably a result of "The great merge" plus cvs to svn to git conversion).

A `CHANGES` entry from back then notes:

> The use of the numinputs argument is reserved for future expansion. Currently, values >1 will generate an error.  However, future releases of SWIG may utilize that to support multi-input typemaps.

However, multiple argument types **are** supported, and I think that's what you want here assuming you want to pass a C# array, which knows its own length:

https://www.swig.org/Doc4.1/Typemaps.html#Typemaps_multi_argument_typemaps

If you are wanting to if you want to handle two parameters from C# together to give two parameters in C++ that may not be cleanly supported and you'll probably need to hack around to achieve it (e.g. by defining and wrapping a `%extend` function which takes a `struct` with the two args to pass to C++, which just forwards to the actual function).
